homeassistant/components/backup/manager.py

```diff
@@ -762,24 +799,40 @@ async def _async_finish_backup(
                     state=CreateBackupState.IN_PROGRESS,
                 )
             )
-            agent_errors = await self._async_upload_backup(
-                backup=written_backup.backup,
-                agent_ids=agent_ids,
-                open_stream=written_backup.open_stream,
-            )
-            await written_backup.release_stream()
-            if with_automatic_settings:
-                # create backup was successful, update last_completed_automatic_backup
-                self.config.data.last_completed_automatic_backup = dt_util.now()
-                self.store.save()
-            self.known_backups.add(written_backup.backup, agent_errors)
 
-            # delete old backups more numerous than copies
-            await delete_backups_exceeding_configured_count(self)
+            try:
+                agent_errors = await self._async_upload_backup(
+                    backup=written_backup.backup,
+                    agent_ids=agent_ids,
+                    open_stream=written_backup.open_stream,
+                )
+                # TODO: Should we track backups that raise due to writer errors?
+                # Should we release the stream in case of writer errors?
```

Comment:
I'm not completely sure what we should do here. I'd like input.

Reply:
I'm a bit confused, because the comment is added where we upload the newly created backup to agents, but it talks about errors from the writer.
Are you talking about errors raised by the `open_stream` method the writer provided?

Reply:
Yes. With the current logic, if there's a writer error during upload, we won't track the backup in known backups and we won't try to release the stream.

Reply:
I see, then I think it's not just about how to handle expected writer errors.

The problem is that we let any error which is not `BackupAgentError` bubble up from `_async_upload_backup`, it could be an expected writer error, or it could be unexpected errors from an agent or the writer.

Maybe the behavior would be better defined if we:
- Collect any `Exception` in `agent_errors`, and only let `BaseException` bubble up - I don't think we should abort upload because of one misbehaving agent?
- Track the backup if at least one agent did not meet issues when uploading, if all agents failed we don't track it
- Always call `release_stream()`

Reply:
The writer error doesn't come from the agent. It comes from the writer. It's likely all agents will face the same error when using `open_stream`.

Reply:
Right now we do this distinction: We don't abort the upload on BackupAgentError. Those are stored in agent errors and tracked together with the backup. For other errors we abort the upload since they are common, so all agents will be affected or they are unknown and we don't know what went wrong.

Reply:
As discussed on Discord, we should also store unexpected `Exception` so one bad agent does not break backup creation. We should let a `BackupReaderWriterError` bubble up and abort backup creation though.